include/os/linux/kernel/linux/dcache_compat.h

```diff
@@ -61,4 +61,25 @@ d_clear_d_op(struct dentry *dentry)
 	    DCACHE_OP_REVALIDATE | DCACHE_OP_DELETE);
 }
 
+/*
+ * Walk and invalidate all dentry aliases of an inode
+ * unless it's a mountpoint
+ */
+static inline void
+zpl_d_drop_aliases(struct inode *inode)
+{
+	struct dentry *dentry;
+	spin_lock(&inode->i_lock);
+#ifdef HAVE_DENTRY_D_U_ALIASES
+	hlist_for_each_entry(dentry, &inode->i_dentry, d_u.d_alias) {
+#else
+	hlist_for_each_entry(dentry, &inode->i_dentry, d_alias) {
+#endif
+		if (!IS_ROOT(dentry) && !d_mountpoint(dentry) &&
+		    (dentry->d_inode == inode)) {
```

Comment:
Wouldn't this inode's dentry list only contain dentries for this inode?

In any case, I agree with @behlendorf that we should remove everything about this filesystem from the directory entry cache, since a `rename` could result in invalid dentry cache even though zfs_rezget() would succeed.

Reply:
> Wouldn't this inode's dentry list only contain dentries for this inode?

Negative dentries used be identified by setting `dentry->d_inode == NULL`.  Though, I'm not certain if they ever were include in the` ->i_dentry`, they may only have been hashed.  Looking at more recent kernels they've add some `DCACHE_ENTRY_TYPE` bits to `->d_flags` since there are now more dentry types.  Regardless, unless there's some reason we can't we should drop everything to force a new lookup.